File: 0661-image-smoother/0661-image-smoother.py
class Solution:
    def imageSmoother(self, img: List[List[int]]) -> List[List[int]]:
        directions = [(0, 1), (1, 1), (1, 0), (1, -1), (0, -1), (-1, -1), (-1, 0), (-1, 1)]

        newImg = []

        for r in range(len(img)):
            newImg.append([])
            for c in range(len(img[r])):
                cells = [img[r][c]]
                for dr, dc in directions:
                    move = (r + dr, c + dc)
                    if (
                        move[0] < 0 or
                        move[0] == len(img) or
                        move[1] < 0 or
                        move[1] == len(img[0])
                    ):
                        continue

                    cells.append(img[move[0]][move[1]])
                
                avg = math.floor(sum(cells) / len(cells))
                newImg[r].append(avg)

        return newImg

File: 0661-image-smoother/0661-image-smoother.cpp
class Solution {
public:
    vector<vector<int>> imageSmoother(vector<vector<int>>& img) {
        vector<vector<int>> resImg = img;
        const vector<pair<int, int>> directions {
            {0, 0},
            {-1, 0},
            {-1, 1},
            {0, 1},
            {1, 1},
            {1, 0},
            {1, -1},
            {0, -1},
            {-1, -1}
        };
        for (int i = 0; i < img.size(); ++i) {
            for (int j = 0; j < img[0].size(); ++j) {
                int numValues = 0;
                int sumTotal = 0;
                for (pair<int, int> dir : directions) {
                    const pair<int, int> move {i + dir.first, j + dir.second};
                    if (isValidCoord(move, img.size(), img[0].size())) {
                        ++numValues;
                        sumTotal += img[i + dir.first][j + dir.second];
                    }
                }
                resImg[i][j] = sumTotal / numValues;
            }
        }
        return resImg;
    }
    bool isValidCoord(const pair<int, int> coord, const int rows, const int cols) {
        if (
            coord.first < 0 ||
            coord.second < 0 ||
            coord.first >= rows ||
            coord.second >= cols
        ) {
            return false;
        }
        return true;
    }
};